client/app/react-components/QueryEditor.js

```diff
@@ -0,0 +1,229 @@
+    table.columns.forEach((c) => {
+      keywords[c] = 'Column';
+      keywords[`${table.name}.${c}`] = 'Column';
+    });
+  });
+  return map(keywords, (v, k) => ({
+    name: k,
+    value: k,
+    score: 0,
+    meta: v,
+  }));
+}
+
+export default class QueryEditor extends React.Component {
+  static propTypes = {
+    queryText: PropTypes.string.isRequired,
+    schema: Schema, // eslint-disable-line react/no-unused-prop-types
+    dataSources: PropTypes.arrayOf(DataSource),
+    dataSource: DataSource,
+    canEdit: PropTypes.bool.isRequired,
+    isDirty: PropTypes.bool.isRequired,
+    isQueryOwner: PropTypes.bool.isRequired,
+    updateDataSource: PropTypes.func.isRequired,
+    canExecuteQuery: PropTypes.func.isRequired,
+    executeQuery: PropTypes.func.isRequired,
+    queryExecuting: PropTypes.bool.isRequired,
+    saveQuery: PropTypes.func.isRequired,
+    updateQuery: PropTypes.func.isRequired,
+    addNewParameter: PropTypes.func.isRequired,
+    listenForResize: PropTypes.func.isRequired,
+    listenForEditorCommand: PropTypes.func.isRequired,
+
+  }
+
+  static defaultProps = {
+    schema: null,
+    dataSource: { options: { doc: '' }, syntax: 'sql' },
+    dataSources: [],
+  }
+
+  constructor(props) {
+    super(props);
+    this.refEditor = React.createRef();
+    this.state = {
+      schema: null,
+      keywords: [],
+    };
+    langTools.addCompleter({
+      getCompletions: (state, session, pos, prefix, callback) => {
+        if (prefix.length === 0) {
+          callback(null, []);
+          return;
+        }
+        callback(null, this.state.keywords);
+      },
+    });
+
+    this.onLoad = (editor) => {
+      // Release Cmd/Ctrl+L to the browser
+      editor.commands.bindKey('Cmd+L', null);
+      editor.commands.bindKey('Ctrl+P', null);
+      editor.commands.bindKey('Ctrl+L', null);
+
+      // eslint-disable-next-line react/prop-types
+      this.props.QuerySnippet.query((snippets) => {
+        const snippetManager = snippetsModule.snippetManager;
+        const m = {
+          snippetText: '',
+        };
+        m.snippets = snippetManager.parseSnippetFile(m.snippetText);
+        snippets.forEach((snippet) => {
+          m.snippets.push(snippet.getSnippet());
+        });
+        snippetManager.register(m.snippets || [], m.scope);
+      });
+      editor.focus();
+      this.props.listenForResize(() => editor.resize());
+      this.props.listenForEditorCommand((e, command, ...args) => {
+        switch (command) {
+          case 'focus': {
+            editor.focus();
+            break;
+          }
+          case 'paste': {
+            const [text] = args;
+            editor.session.doc.replace(editor.selection.getRange(), text);
+            const range = editor.selection.getRange();
+            this.props.updateQuery(editor.session.getValue());
+            window.setTimeout(() => {
+              editor.selection.setRange(range);
+            });
```

Comment:
If I'm not mistaken, [we're using $timeout in the Angular component](https://github.com/getredash/redash/blob/master/client/app/components/queries/query-editor.js#L57-L59) to sync with Angular's digest loop. Not sure if we need this here. But maybe @kravets-levko can shed more light?

Reply:
@arikfr Actually, I don't remember why we need `$timeout` here :-( Probably, there were some issues when trying to restore selection, and `$timeout` fixed that, but it's better just to try - remove `window.setTimeout` and check if it works correctly.